minecode_pipelines/miners/swift.py

```diff
@@ -0,0 +1,47 @@
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# purldb is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/aboutcode-org/purldb for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+import shutil
+import subprocess
+from urllib.parse import urlparse
+
+
+def is_safe_repo_url(repo_url: str) -> bool:
+    parsed = urlparse(repo_url)
+    return (
+        parsed.scheme == "https"
+        and parsed.netloc == "github.com"
+        and parsed.path.endswith(".git")
+        or parsed.path.count("/") >= 2
+    )
+
+
+def fetch_tags_raw(repo_url: str, timeout: int = 60, logger=None) -> str | None:
+    """Run `git ls-remote` on a GitHub repo and return raw output, or None on error."""
+    git_executable = shutil.which("git")
+    if git_executable is None:
+        logger("Git executable not found in PATH")
+        return None
+
+    if not is_safe_repo_url(repo_url):
+        raise ValueError(f"Unsafe repo URL: {repo_url}")
+
+    try:
+        result = subprocess.run(
+            [git_executable, "ls-remote", repo_url],
+            capture_output=True,
+            text=True,
+            check=True,
+            timeout=timeout,
```

Comment:
I’m not sure what the best approach is to avoid any possibility of OS command execution.

Reply:
@ziadhany this works, and I was tempted to use GitPython, but this only does a spawn a subprocess too:
- https://github.com/gitpython-developers/GitPython/blob/2d4c54154bc984f5dafde1b23e377e2dc3214cb8/test/lib/helper.py#L307

You will need to have a single, shared function to list tags using ls-remote in the future to avoid duplication with https://github.com/aboutcode-org/purldb/blob/c468a762a13286e857ca236ebf7dadb5be157c57/purl2vcs/src/purl2vcs/find_source_repo.py#L384

Just create an issue with this comment to track for later.

See also for reference:
- https://github.com/aboutcode-org/vulnerablecode/issues/493